### DISCUSSION TITLE
Can STL be used?

### DISCUSSION THREAD

**Discussion by ConstStar** [nasa/fprime#2567] [Q&A] [answered]
Why can't fprime use STL? If my device has sufficient performance, can I use STL?
Because some of the third-party libraries I have introduced use STL and have not found any alternatives such as std:: vector or std:: map.
Thanks

**Comment by LeStarch** [comment, marked as answer]
Like many shiny c++ features, a project may absolutely choose to use the standard template library (STL).  It is absolutely required to work with some third-party code.  

In F´ core, we don't allow STL simply because this code is used by many different projects some, which might choose to use STL and others that do not.  The reasoning behind this is because the STL uses dynamic memory management to allocate memory for the object stored in things like the `std::vector` and this runtime memory allocation violates the "no allocation after initialization" rule. Hence many projects are wary of the STL.

There are ways to offset these allocations.  A simple method would be to preallocate sizes for STL construct ... but STL may reallocate those anyway resulting in more memory allocation.  Alternatively, one could create an STL-allocator implementation that delegates to an F´ buffer pool (BufferManager) but this hasn't been proved before.

In all things, your project should trade the risks of using the STL with the benefits of using.  Let me know if you would like some elaboration.

**Reply by ConstStar** [reply]
Is it just that STL is not recommended for embedded systems in general, or does Fprime have some impact on using STL? If I don’t use STL as a member variable in Fprime’s classes that need serialization, but only as a local variable in member functions, and my device has enough memory, such as 2G or 4G, can I use STL?

**Reply by LeStarch** [reply]
I would not use STL objects as arguments to ports or commands, but otherwise F´ does not prevent users from using these things.  In general embedded systems tend to be wary of the memory implications of the STL and project implementers should understand the tradespace before using such things.

**Reply by ConstStar** [reply]
I understand now, thank you for explaining.
